src/ndarray/ndarray.cc

```diff
@@ -687,53 +687,57 @@ void NDArray::SyncCopyFromCPU(const void *data, size_t size) const {
   TShape dshape = this->shape();
   CHECK_EQ(dshape.Size(), size)
       << "Memory size do not match";
-  Context ctx = this->ctx();
   TBlob dst = this->data();
   TBlob src((void*)data, dshape, cpu::kDevMask, this->dtype_); // NOLINT(*)
 
-  RunContext run_ctx;
-  run_ctx.stream = nullptr;
-  if (ctx.dev_mask() == cpu::kDevMask) {
-    ndarray::Copy<cpu, cpu>(src, &dst, Context::CPU(), ctx, run_ctx);
+  if (this->ctx().dev_mask() == cpu::kDevMask) {
+    Engine::Get()->PushSync([&](RunContext rctx) {
+        ndarray::Copy<cpu, cpu>(src, &dst,
+                                Context::CPU(), Context::CPU(), rctx);
+      }, this->ctx(), {}, {this->var()},
+      FnProperty::kNormal, 0, PROFILER_MESSAGE("SyncCopyCPU2CPU"));
   } else {
 #if MXNET_USE_CUDA
-    // use empty stream to do sync copy
-    // TODO(bing, yutian) consider use a Real Stream, so it is not blocking others
-    // Maybe move to engine part
-    mshadow::Stream<gpu> zero_stream;
-    run_ctx.stream = &zero_stream;
-    ndarray::Copy<cpu, gpu>(src, &dst, Context::CPU(), ctx, run_ctx);
+    Engine::Get()->PushSync([&](RunContext rctx) {
+        ndarray::Copy<cpu, gpu>(src, &dst,
+                                Context::CPU(), this->ctx(), rctx);
+        // Wait GPU kernel to complete
+        rctx.get_stream<gpu>()->Wait();
+      }, this->ctx(), {}, {this->var()},
+      FnProperty::kCopyToGPU, 0, PROFILER_MESSAGE("SyncCopyCPU2GPU"));
 #else
     LOG(FATAL) << "GPU is not enabled";
 #endif
   }
+  this->WaitToRead();
 }
 
 void NDArray::SyncCopyToCPU(void *data, size_t size) const {
-  this->WaitToRead();
   TShape dshape = this->shape();
   CHECK_EQ(dshape.Size(), size)
       << "Memory size do not match";
-  Context ctx = this->ctx();
-  TBlob src = this->data();
   TBlob dst(data, dshape, cpu::kDevMask, this->dtype_); // NOLINT(*)
 
-  RunContext run_ctx;
-  run_ctx.stream = nullptr;
-  if (ctx.dev_mask() == cpu::kDevMask) {
-    ndarray::Copy<cpu, cpu>(src, &dst, ctx, Context::CPU(), run_ctx);
+  if (this->ctx().dev_mask() == cpu::kDevMask) {
+    Engine::Get()->PushSync([&](RunContext rctx) {
+        ndarray::Copy<cpu, cpu>(this->data(), &dst,
+                                Context::CPU(), Context::CPU(), rctx);
```

Comment:
Calling from same thread can improve cache locality and removes overhead of calling into engine.